src/daft-micropartition/src/micropartition.rs

```diff
@@ -0,0 +1,219 @@
+struct MicroPartition {
+    schema: SchemaRef,
+    state: Mutex<TableState>,
+    statistics: Option<TableStatistics>,
+}
+
+impl MicroPartition {
+    pub fn new(schema: SchemaRef, state: TableState, statistics: Option<TableStatistics>) -> Self {
+        MicroPartition {
+            schema,
+            state: Mutex::new(state),
+            statistics,
+        }
+    }
+
+    pub fn empty() -> Self {
+        Self::new(
+            Schema::empty().into(),
+            TableState::Loaded(Arc::new(vec![])),
+            None,
+        )
+    }
+
+    fn tables_or_read(&self) -> crate::Result<Arc<Vec<Table>>> {
+        let mut guard = self.state.lock().unwrap();
+
+        match guard.deref() {
+            TableState::Loaded(tables) => Ok(tables.clone()),
+            TableState::Unloaded(params) => {
+                let table_values: Vec<_> = match &params.format_params {
+                    FormatParams::Parquet(parquet_schema_inference) => {
+                        let io_client = daft_io::get_io_client(
+                            params.multithreaded_io,
+                            params.io_config.clone(),
+                        )
+                        .unwrap();
+                        let column_names = params
+                            .columns
+                            .as_ref()
+                            .map(|v| v.iter().map(|s| s.as_ref()).collect::<Vec<_>>());
+                        let urls = params.urls.iter().map(|s| s.as_str()).collect::<Vec<_>>();
+                        daft_parquet::read::read_parquet_bulk(
+                            urls.as_slice(),
+                            column_names.as_deref(),
+                            None,
+                            params.limit,
+                            None,
+                            io_client.clone(),
+                            8,
+                            params.multithreaded_io,
+                            parquet_schema_inference,
+                        )
+                        .context(DaftCoreComputeSnafu)?
+                    }
+                };
+
+                let filter_preds = &params.filters;
+                let table_values = if !filter_preds.is_empty() {
+                    table_values
+                        .into_iter()
+                        .map(|t| t.filter(filter_preds.as_slice()))
+                        .collect::<Result<Vec<Table>, _>>()
+                        .context(DaftCoreComputeSnafu)?
+                } else {
+                    table_values
+                };
+
+                let table_values = Arc::new(table_values);
+                *guard = TableState::Loaded(table_values.clone());
+                Ok(table_values)
+            }
+        }
+    }
+
+    pub fn filter(&self, predicate: &[Expr]) -> super::Result<Self> {
+        if predicate.is_empty() {
+            return Ok(Self::new(
+                self.schema.clone(),
+                TableState::Loaded(vec![].into()),
```

Comment:
Why is empty predicate a full filter instead of a no-op?

E.g. if we perform some predicate pushdown and somehow end up with a `filter([])`, shouldn't that be a no-op?